pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/ConsumptionDelayTracker.java

```diff
@@ -0,0 +1,295 @@
+  private static final int PARTITION_TIMEOUT_MS = 600000;          // 10 minutes timeouts
+  // Delay Timer thread for this time after starting timer
+  private static final int INITIAL_TIMER_THREAD_DELAY_MS = 1000;
+  // Mutable versions of the above constants so we can test with smaller delays
+  final int _timerThreadTickIntervalMs;
+  final int _initialTimeThreadDelayMs;
+  // Timer task to check partitions that are inactive against ideal state.
+  final private Timer _timer;
+
+  final private ServerMetrics _serverMetrics;
+  final private String _tableNameWithType;
+
+  private boolean _enableAging;
+  final private Logger _logger;
+
+  final private RealtimeTableDataManager _realTimeTableDataManager;
+
+  /*
+   * Helper function to update the maximum when the current maximum is removed or updated.
+   * If no samples left we set maximum to minimum so new samples can be recorded.
+   */
+  private DelayMeasure getMaximumDelay() {
+    DelayMeasure newMax = null;
+    for (long pid : _partitionToDelaySampleMap.keySet()) {
+      DelayMeasure currentMeasure = _partitionToDelaySampleMap.get(pid);
+      if ((newMax == null) || (currentMeasure._delayMilliseconds > newMax._delayMilliseconds)) {
+        newMax = currentMeasure;
+      }
+    }
+    return newMax;
+  }
+
+  private List<Integer> getPartitionsHostedByThisServerPerIdealState() {
+    return _realTimeTableDataManager.getTablePartitionsHostedByThisServer();
+  }
+  /*
+   * Helper function to be called when we should stop tracking a given partition. Removes the partition from
+   * all our maps, it also updates the maximum if the tracked partition was the previous maximum.
+   *
+   * @param partitionGroupId partition ID which we should stop tracking.
+   */
+  private void removePartitionId(long partitionGroupId) {
+    _partitionToDelaySampleMap.remove(partitionGroupId);
+    // If we are removing a partition we should stop reading its ideal state.
+    _partitionsMarkedForVerification.remove(partitionGroupId);
+   }
+
+  // Custom Constructor
+  public ConsumptionDelayTracker(ServerMetrics serverMetrics, String tableNameWithType,
+      RealtimeTableDataManager realtimeTableDataManager, int initialTimeThreadDelayMs, int timerThreadTickIntervalMs)
+      throws RuntimeException {
+    _logger = LoggerFactory.getLogger(tableNameWithType + "-" + getClass().getSimpleName());
+    _serverMetrics = serverMetrics;
+    _tableNameWithType = tableNameWithType;
+    _realTimeTableDataManager = realtimeTableDataManager;
+    // Handle negative timer values
+    if ((initialTimeThreadDelayMs < 0) || (timerThreadTickIntervalMs <= 0)) {
+      throw new RuntimeException("Illegal timer arguments");
+    }
+    _enableAging = true;
+    _initialTimeThreadDelayMs = initialTimeThreadDelayMs;
+    _timerThreadTickIntervalMs = timerThreadTickIntervalMs;
+    _timer = new Timer("ConsumptionDelayTimerThread" + tableNameWithType);
```

Comment:
raw table name should be good enough in this line as well as next

Reply:
I think this is better for debug purposes, leaving as is

Reply:
same as before, Long names show up in thread dumps, etc. We should be ok with raw table name, please remove, thanks